src/test/java/picard/sam/markduplicates/UmiAwareMarkDuplicatesWithMateCigarTest.java

```diff
@@ -0,0 +1,155 @@
+                        Arrays.asList(new Boolean[] {false, true, true, true, true}),
+                        2
+                },
+                {
+                        // Test basic error correction using edit distance of 1 where UMIs
+                        // form a chain in edit distance space so that a UMI with large
+                        // edit distance will get error corrected to a distant but linked (in edit space) UMI
+                        Arrays.asList(new String[] {"AAAA", "AAAA", "AAAT", "AAGT", "ACGT", "TCGT", "CCCC"}),
+                        Arrays.asList(new String[] {"AAAA", "AAAA", "AAAA", "AAAA", "AAAA", "AAAA", "CCCC"}),
+                        Arrays.asList(new Boolean[] {false, true, true, true, true, true, false}),
+                        1
+                },
+                {
+                        // Test short UMIs
+                        Arrays.asList(new String[] {"A", "A", "T", "G", "G", "C", "C", "A"}),
+                        Arrays.asList(new String[] {"A", "A", "A", "A", "A", "A", "A", "A"}), // All UMIs should get corrected to A
+                        Arrays.asList(new Boolean[] {false, true, true, true, true, true, true, true}), // All mate pairs should be duplicates except the first
+                        1
+                },
+                {
+                        // Test short UMIs with no allowance for errors
+                        Arrays.asList(new String[] {"A", "A", "T", "G", "G", "C", "C", "A"}),
+                        Arrays.asList(new String[] {"A", "A", "T", "G", "G", "C", "C", "A"}), // No UMIs should get corrected
+                        Arrays.asList(new Boolean[] {false, true, false, false, true, false, true, true}), // Only exactly duplicated UMIs will give rise to a new duplicate set
+                        0
+                },
+                {
+                        // Test longish UMIs with relatively large allowance for error
+                        // UMIs "TTGACATCCA", "ATGCCATCGA", "AAGTCACCGT" should belong to the same duplicate set since
+                        // they are within edit distance of 4 of each other.  TTGACATCCA should be chosen as the inferred
+                        // UMI even though it only occurs once.  Since all UMIs only occur once, we choose the UMI that
+                        // is not marked as duplicate to be the inferred UMI.
+                        Arrays.asList(new String[] {"TTGACATCCA", "ATGCCATCGA", "AAGTCACCGT"}),
+                        Arrays.asList(new String[] {"TTGACATCCA", "TTGACATCCA", "TTGACATCCA"}), // All UMIs should get corrected to TTGACATCCA
+                        Arrays.asList(new Boolean[] {false, true, true}), // All mate pairs should be duplicates except the first
+                        4
+                },
+                {
+                        // Test to make sure that if any reads don't have a UMI, we treat things as if there were no
+                        // UMIs at all
```

Comment:
I think an command line option `ALLOW_MISSING_UMIS` is fine, though I still would like a practical reason for it.  Could you give me an example where such an option would be useful?